pytket/pytket/rmbqc/knitting.py

```diff
@@ -0,0 +1,331 @@
+# Copyright 2019-2022 Cambridge Quantum Computing
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from enum import Enum
+from pytket.mapping import get_token_swapping_network  # type: ignore
+from pytket.circuit import Circuit, Bit, Qubit, OpType  # type: ignore
+from pytket.predicates import CompilationUnit  # type: ignore
+from pytket.passes import RoutingPass  # type: ignore
+from pytket.architecture import Architecture, FullyConnected  # type: ignore
+from typing import Dict, Tuple, List, Union
+
+
+class Knitter(Enum):
+    @staticmethod
+    def separate(
+        pattern_list: List[Tuple[Circuit, Dict[str, Dict[Qubit, int]]]],
+        arch: Architecture,
+    ) -> Tuple[Circuit, Dict[str, Dict[Qubit, int]]]:
+        """
+        Takes a list of tuples containing pytket circuits and their corresponding
+        input/output maps as well as an architecture. It then routes each circuit on the
+        architecture separately. Finally it connects the outputs of one segment to
+        the inputs of the next by generating a network of SWAP gates. Then returns
+        a tuple containing the new circuit object and its corresponding input/output map.
+        
+        :param pattern_list:  List of tuples of circuits with their input/output maps.
+                                These map the input/output qubits of the original circuit
+                                to their new locations in the qubit register.
+        :param type:          List[Tuple[Circuit,Dict[str, Dict[Qubit, int]]]]
+        
+        :param arch:          The architecture onto which to fit the circuits.
+        :param type:          Architecture
+        
+        :returns:        A tuple containing a circuit and an input/output map dictionary.
+        :rtype:          Tuple[Circuit,Dict[str, Dict[Qubit, int]]]
+        """
+        new_c = Circuit()
+        for q in arch.nodes:
+            new_c.add_qubit(q)
+        for p in range(len(pattern_list)):
+            (curr_circ, curr_map) = pattern_list[p]
+            route_map = {}
+            if p == 0:
```

Comment:
rather than having a p==0 condition, please can you do this outside the loop and then do the from p = 1

Reply:
I think I did it like this so that it doesn't fail for an empty circuit... If you have an empty circuit it will be converted into 0 segments so it's not gonna go through any iterations of the loop.